packages/truffle-decode-utils/src/types/values.ts

```diff
@@ -0,0 +1,1423 @@
+    message() {
+      let typeName = Types.isContractDefinedType(this.type)
+        ? this.type.definingContractName + "." + this.type.typeName
+        : this.type.typeName;
+      return `Unknown ${this.type.typeClass} type ${typeName} of id ${this.type.id}`;
+    }
+    constructor(unknownType: Types.UserDefinedType) {
+      super();
+      this.type = unknownType;
+      this.kind = "UserDefinedTypeNotFoundError";
+    }
+  }
+
+  //Read errors
+  export class UnsupportedConstantError extends GenericError {
+    kind: "UnsupportedConstantError";
+    definition: AstDefinition;
+    message() {
+      return `Unsupported constant type ${DefinitionUtils.typeClass(this.definition)}$`;
+    }
+    constructor(definition: AstDefinition) {
+      super();
+      this.definition = definition;
+      this.kind = "UnsupportedConstantError";
+    }
+  }
+
+  export class ReadErrorStack extends GenericError {
+    kind: "ReadErrorStack";
+    from: number;
+    to: number;
+    message() {
+      return `Can't read stack from position ${this.from} to ${this.to}`;
+    }
+    constructor(from: number, to: number) {
+      super();
+      this.from = from;
+      this.to = to;
+      this.kind = "ReadErrorStack";
+    }
+  }
+
+  //finally, a convenience function for constructing generic errors
+  export function makeGenericErrorResult(dataType: Types.Type, error: GenericError): ErrorResult {
+    switch(dataType.typeClass) {
+      case "uint":
+        return new UintErrorResult(dataType, error);
+      case "int":
+        return new IntErrorResult(dataType, error);
+      case "bool":
+        return new BoolErrorResult(dataType, error);
+      case "bytes":
+        switch(dataType.kind) {
+          case "static":
+            return new BytesStaticErrorResult(dataType, error);
+          case "dynamic":
+            return new BytesDynamicErrorResult(dataType, error);
+        }
+      case "address":
+        return new AddressErrorResult(dataType, error);
+      case "fixed":
+        return new FixedErrorResult(dataType, error);
+      case "ufixed":
+        return new UfixedErrorResult(dataType, error);
+      case "string":
+        return new StringErrorResult(dataType, error);
+      case "array":
+        return new ArrayErrorResult(dataType, error);
+      case "mapping":
+        return new MappingErrorResult(dataType, error);
+      case "struct":
+        return new StructErrorResult(dataType, error);
+      case "enum":
+        return new EnumErrorResult(dataType, error);
+      case "contract":
+        return new ContractErrorResult(dataType, error);
+      case "magic":
+        return new MagicErrorResult(dataType, error);
+      case "function":
+        switch(dataType.visibility) {
+          case "external":
+            return new FunctionExternalErrorResult(dataType, error);
+          case "internal":
+            return new FunctionInternalErrorResult(dataType, error);
+        }
+    }
+  }
+}
```

Comment:
We've talked about this a lot, so I don't think it's worth reiterating the challenges here. Suffice to say, the current of state of Values here is definitely sufficient for this implementation, given all the necessary constraints with TS.

Nice job fighting the type system. I just hope they continue to make that fight easier.